sdks/python/apache_beam/runners/portability/fn_api_runner.py

```diff
@@ -277,6 +277,159 @@ def deduplicate_read(self):
 
     safe_coders = {}
 
+    def lift_combiners(stages):
+      """Expands CombinePerKey into pre- and post-grouping stages.
+
+      ... -> CombinePerKey -> ...
+
+      becomes
+
+      ... -> PreCombine -> GBK -> MergeAccumulators -> ExtractOutput -> ...
+      """
+      def new_id(existing_ids, prefix=''):
+        existing_ids = set(existing_ids)
+        counter = 0
+        while True:
+          new_id = '%s_%d' % (prefix, counter)
+          if new_id not in existing_ids:
+            return new_id
+          counter += 1
+
+      def add_or_get_coder_id(coder_proto):
+        for coder_id, coder in pipeline_components.coders.items():
+          if coder == coder_proto:
+            return coder_id
+        new_coder_id = new_id(pipeline_components.coders, 'coder')
+        pipeline_components.coders[new_coder_id].CopyFrom(coder_proto)
+        return new_coder_id
+
+      def windowed_coder_id(coder_id):
+        proto = beam_runner_api_pb2.Coder(
+            spec=beam_runner_api_pb2.SdkFunctionSpec(
+                spec=beam_runner_api_pb2.FunctionSpec(
+                    urn=urns.WINDOWED_VALUE_CODER)),
+            component_coder_ids=[coder_id, window_coder_id])
+        return add_or_get_coder_id(proto)
+
+      for stage in stages:
+        assert len(stage.transforms) == 1
+        transform = stage.transforms[0]
+        if transform.spec.urn == urns.COMBINE_PER_KEY_TRANSFORM:
+          combine_payload = proto_utils.parse_Bytes(
+              transform.spec.payload, beam_runner_api_pb2.CombinePayload)
+
+          input_pcoll = pipeline_components.pcollections[only_element(
+              transform.inputs.values())]
```

Comment:
<!--new_thread; commit:15c1178c61218f4f56ad3d74a1d52f3f1dc6e727; resolved:0-->
Does transform.inputs.values() always return one value?

Reply:
It does for COMBINE_PER_KEY_TRANSFORM.